fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/impl/transfer/item/ItemVariantImpl.java

```diff
@@ -48,6 +49,10 @@ public static ItemVariant of(Item item, @Nullable NbtCompound tag) {
 	private final Item item;
 	private final @Nullable NbtCompound nbt;
 	private final int hashCode;
+	/**
+	 * Lazily computed, equivalent to calling toStack(1). <b>MAKE SURE IT IS NEVER MODIFIED!</b>
+	 */
+	private volatile @Nullable ItemStack cachedStack = null;
```

Comment:
```suggestion
	private @Nullable ItemStack cachedStack;
```
Since you already have the data race guard in `getCachedStack()`. `volatile` is expensive.

Reply:
I think it's still needed since there is no guarantee that the ItemStack reference will be set and properly initialized _after_ the ItemStack is fully constructed.

Reply:
Maybe @sfPlayer1 has a better suggestion here? Would the following be correct?
```java
private @Nullable ItemStack cachedStack = null;
public ItemStack getCachedStack() {
	ItemStack ret = cachedStack;

	if (ret == null) {
		ret = toStack();
		VarHandle.fullFence(); // Ensure stack is fully constructed before it's assigned to cachedStack
		cachedStack = ret;
	}

	return ret;
}
```

Reply:
@Technici4n Why can't the cached stack be constructed twice? Is there a special requirement here that the `cachedStack` must be the same reference than equivalent objects?

Reply:
Hmm, I think you can probably use a `storeStoreFence` instead: https://github.com/openjdk/jdk/blob/5a768f75c9cb013edbf6c61e79820bd180cad4ba/test/micro/org/openjdk/bench/vm/fences/SafePublishing.java#L76

Reply:
@Technici4n Still need to fix this. Using a full fence is still better than volatile, which slows down every read.

Reply:
Yeah, I'm just not sure it's completely correct. 😅

Reply:
No idea, I would like Player to have a look if we decide to go for a fence. But I think keeping volatile for now should be fine performance-wise.

Reply:
The recommended VarHandle pattern btw is to use volatile on the field and then do lesser accesses through handles as appropriate

Reply:
I am no expert on this, but my understanding is that the suggested fence implementation probably doesn't work?
````
private @Nullable ItemStack cachedStack = null;
public ItemStack getCachedStack() {
	ItemStack ret = cachedStack; // R1

	if (ret == null) {
		ret = toStack();
		VarHandle.fullFence(); // Ensure stack is fully constructed before it's assigned to cachedStack
		cachedStack = ret; // R2
	}

	return ret;
}
````
The issue is that although the ````fullFence()```` will flush everything to the cpu caches and main memory which seems to be same as volatile, the main contract in the java memory is a "happens before" that the compiler can see/calculate to generate correctly ordered code. This exists for R2 above, but not R1. While it does exist at R1 if cachedStack is volatile.

Reply:
I believe the java fence stuff is just a passthrough of the c++ stuff?
https://en.cppreference.com/w/cpp/atomic/atomic_thread_fence if you want to check the rules.
Basically for it work, it needs a synchronize, park, volatile r/w, etc. somewhere on the reader thread competing with the constucting thread which is not true without the volatile on the cachedStack.

Reply:
This is only to ensure if the `cachedStack` is read to be `non-null` in a benign-race, we can see the nbt data (non-final) written to the stack object. We don't require `cachedStack` to be always visible after it is written (like volatiles) as another thread can just create another cached stack on-demand; but when one thread's cache is seen by another thread, we require its nbt data (non-final, prepared by `toStack()`) to be seen as well, which is what the fence does.

Reply:
````
but when one thread's cache is seen by another thread, we require its nbt data (non-final, prepared by toStack()) to be seen as well, which is what the fence does
````
But I don't think it does, except when you have other heuristics like x86's memory writes are always globally visual once flushed (not true generally in "weaker" architectures).
The issue is that without the "synchronization" somewhere in the reading thread there's no guarantee it runs the loads in the correct order, or even that it does any loads at all, not even from the cpu cache - the compiler is within its rights to generate code that uses established but "out-of-date" values in registers or the cpu uses a different cache/layer.

Maybe you are correct that in this case these bad things can't happen? 

I believe there would be less of a problem if the cache was immutable/static final because the spec requires the compiler to go the extra mile to correctly order and completly construct things, but that is only relevant for things like lazy immutable singletons.

Reply:
I'm really not sure, so we should probably just keep the volatile. Acquire/release might work, but volatile is a lot easier to write, maintain, and reason about, so I suggest keeping it for now.

Reply:
I think a store-store fence is enough to create the final semantics. https://github.com/openjdk/jdk/blob/9583e3657e43cc1c6f2101a64534564db2a9bd84/src/java.base/share/classes/java/lang/invoke/ConstantCallSite.java#L52

Notice our fence applies between the write to stack content and the write to `cachedStack` field. On the reader side, we always have that if the stack nbt is read, `cachedStack` pointer is already there.

Since synchronization comes at such a great cost, then I'd recommend not cache unless the stack has complex nbt data if you think fences are unacceptable.

Reply:
> I'm really not sure, so we should probably just keep the volatile. Acquire/release might work, but volatile is a lot easier to write, maintain, and reason about, so I suggest keeping it for now.

I'd suggest removing caching once and for all. Making a single item stack (especially most item keys don't have nbt) would be cheaper than synchronizing.

Reply:
Hmm well, we also have the option of using an ItemStack as the backing storage of course.

Reply:
> Hmm well, we also have the option of using an ItemStack as the backing storage of course.

Uh, I mean to roll back all the changes of `toStack` to `getCachedStack` back to plain `toStack()`.

Reply:
That doesn't make sense imo, I ran some tests locally and it vastly improved the performance in one of my tests. I don't think volatile is that slow, after all it's used for events.

Reply:
(Or even API Lookup)

Reply:
We can keep this for now. I can write a jcstress test later